### ISSUE TITLE
Dynamic number of handle for custom node

### ISSUE THREAD

**Issue by Ipicky22** [wbkd/react-flow#1641]
Hi,

I try to custom a node, which would contain a + and - button and would add or remove a "connection ring".
I try something, the first connection ring appears, but not the following. And when I try to connect the first, I have an error.

First I show you my custom node.
https://zupimages.net/viewer.php?id=21/44/30yf.png

CustomNode :

```Javascript
import React, { memo, useState } from 'react';
import { Handle, Position } from 'react-flow-renderer';
import Checkbox from '@mui/material/Checkbox';
import ButtonGroup from '@mui/material/ButtonGroup/ButtonGroup';
import Button from '@mui/material/Button/Button';

const IncNode = memo(({ isConnectable, sourcePosition, targetPosition, data, ...props }: any) => {

    const [targetArray, setTargetArray] = useState<any>([])

    const add = (type: string) => {
        let tmp = type + (targetArray.length + 1)
        setTargetArray([...targetArray, tmp])
    }

    return (
        <>
            {targetArray && targetArray.map((id: string, key: any) => (
                <Handle
                    key={key}
                    type="target"
                    position={targetPosition ? targetPosition : Position.Left}
                    id={id}
                    style={{ background: '#555' }}
                    isConnectable={isConnectable}
                />
            ))}

            <div style={{ display: "flex", justifyContent: "space-around", alignItems: "center" }}>
                <div style={{ flex: 1 }}>
                    <ButtonGroup
                        orientation="vertical"
                        aria-label="vertical outlined button group"
                        size="small"
                    >
                        <Button key="targetMore" onClick={() => add("T")}>+</Button>
                    </ButtonGroup>
                </div>
                <div style={{ flex: 2 }}>
                    <div style={{ fontWeight: 500, fontSize: 15 }}>{data.label}</div>
                </div>
        </>
    );
});

export default IncNode
```

I simplified the code by removing the icons, the less function, and the buttons on the right.
When I try to connect the first connection ring I have this error : 

```Console
The above error occurred in the <ConnectionLine> component:
    at ConnectionLine


cjs.js:12 Uncaught TypeError: Cannot read properties of null (reading 'find')
    at ConnectionLine (cjs.js:12)
```

How can I increase my number of connection rings, and that they work when I want to connect them ?

Thank you in advance for your answers

**Comment by mermicode** [comment]
Hey, if I have understood your point correctly with "connection ring" you mean the handles right?

If so, there's an example you can use. Basically, you need to create a function for your handles and pass in the number of handles you want and a value to calculate the spacing (that is up to you and you need to make sure they don't overlap) them a look at this from the examples:

```javascript
import React, { memo, FC, useMemo, CSSProperties } from 'react';

import { Handle, Position, NodeProps } from 'react-flow-renderer';

const nodeStyles: CSSProperties = { padding: 10, border: '1px solid #ddd' };

const CustomNode: FC<NodeProps> = ({ data }) => {
  const handles = useMemo(
    () =>
      Array.from({ length: data.handleCount }, (x, i) => {
        const handleId = `handle-${i}`;
        return (
          <Handle
            key={handleId}
            type="source"
            position={Position.Right}
            id={handleId}
            style={{ top: 10 * i + data.handlePosition * 10 }}
          />
        );
      }),
    [data.handleCount, data.handlePosition]
  );

  return (
    <div style={nodeStyles}>
      <Handle type="target" position={Position.Left} />
      <div>output handle count: {data.handleCount}</div>
      {handles}
    </div>
  );
};

export default memo(CustomNode);
```
Finally your error, I've come across many of these so I can't remember the exact one but I would always recommend anyone dealing with connecting nodes check the following:
1. It has to be a source to a target or the opposite
2. the nodes need to have ID as I think these are the ones in the edge data
3. I you have multiple handles, you need to give them handle ID's as the snipper above ('id: string - you only need this when you have multiple source or target handles (otherwise the node id is used)' - https://reactflow.dev/docs/api/handle/)

**Comment by Ipicky22** [comment]
Hey ! Thanks a lot for the answer !

After a few hours of code I almost got what I wanted.

![CustomNode](https://user-images.githubusercontent.com/37102040/140491978-979b683a-fef1-449c-b3e2-6fae9b85b4b0.gif)

I will explain how I did it, because I think there are several things that ReactFlow users may be interested in.

```Javascript
const IncNode = ({ isConnectable, sourcePosition, targetPosition, data, ...props }: any) => {
    const nodeRef: any = useRef();
    const [targetArray, setTargetArray] = useState<any>([])
    const [dimensions, setDimensions] = useState({ width: 20, height: 20 });

    useLayoutEffect(() => {
        if (nodeRef.current) {
            setDimensions({
                width: nodeRef.current.offsetWidth + dimensions.width,
                height: nodeRef.current.offsetHeight + dimensions.height
            });
        }
    }, []);

    const add = (type: string) => {
        if (type === "T" && targetArray.length < 4) {
            let tmp = (targetArray.length + 1)
            setTargetArray([...targetArray, tmp])
        }
    }

    const positionHandle = (index: number) => {
        if (index === 1 || index === 2) {
            return (dimensions.height / 3) * index
        } else if (index === 3) {
            return 0
        } else if (index === 4) {
            return dimensions.height
        }
    }

    const targetHandles = useMemo(
        () =>
            targetArray.map((x: any, i: number) => {
                const handleId = `target-handle-${i + 1}`;
                return (
                    <Handle
                        key={handleId}
                        type="target"
                        position={Position.Left}
                        id={handleId}
                        style={{ top: positionHandle(i + 1) }}
                    />
                );
            }),
        [targetArray]
    );


    return (
        <div ref={nodeRef} >
            {targetHandles}
            <div style={{ display: "flex", justifyContent: "space-around", alignItems: "center" }}>
                <div style={{ flex: 1 }}>
                    <ButtonGroup
                        orientation="vertical"
                        aria-label="vertical outlined button group"
                        size="small"
                    >
                        <Button key="targetMore" onClick={() => add("T")}>+</Button>
                    </ButtonGroup>
                </div>
                <div style={{ flex: 2 }}>
                    <div style={{ fontWeight: 500, fontSize: 15 }}>{data.label}</div>
                </div>
            </div>
        </div>
    );
};

export default memo(IncNode);
```

First I create function for handle, like you said, but I don't have number of handle because I want it dynamically.
And I distinguished source & target by adding it in the ID


Handle Function
```Javascript
const targetHandles = useMemo(
        () =>
            targetArray.map((x: any, i: number) => {
                const handleId = `target-handle-${i + 1}`;
                return (
                    <Handle
                        key={handleId}
                        type="target"
                        position={Position.Left}
                        id={handleId}
                        style={{ top: positionHandle(i + 1) }}
                    />
                );
            }),
        [targetArray]
    );
```

But I had a problem with the placement of the handles, which was not very aesthetic.
So I recovered the size of my node thanks to a ref, to which I added the 20px of padding. Then I did a position assignment function.


Ref
```Javascript
    const nodeRef: any = useRef();
    const [dimensions, setDimensions] = useState({ width: 20, height: 20 });

    useLayoutEffect(() => {
        if (nodeRef.current) {
            setDimensions({
                width: nodeRef.current.offsetWidth + dimensions.width,
                height: nodeRef.current.offsetHeight + dimensions.height
            });
        }
    }, []);

......

<div ref={nodeRef} >
```


Position Assignment Function.
```Javascript
 const positionHandle = (index: number) => {
        if (index === 1 || index === 2) {
            return (dimensions.height / 3) * index
        } else if (index === 3) {
            return 0
        } else if (index === 4) {
            return dimensions.height
        }
    }

.....

style={{ top: positionHandle(i + 1) }}
```

But when I try to connect my handles I always get the same error messages.

```Console
The above error occurred in the <ConnectionLine> component:
    at ConnectionLine


cjs.js:12 Uncaught TypeError: Cannot read properties of null (reading 'find')
    at ConnectionLine (cjs.js:12)
```

So if there is more explanation on how to fix this, I'm fine.

@mermicode, you said "It has to be a source to a target or the opposite"
I admit to being a little lost with this concept in the bookstore. Because for me the source corresponds to the entrance, therefore to the left or above. And the target corresponds to the place where the information will go, the output, so on the right or at the bottom. But in the ReactFlow examples it's the other way around.

**Comment by mermicode** [comment]
@Ipicky22 good stuff mate, so basically what I meant and don't think I see in your code is that handles have types and this can be either source or target. By default you can only connect a source to a target (or target to source, but the direction will be the same) if you try to connect a source to a source or target to target from what I'm reading in your code that's not going to work. (unless there is something else going wrong. So I would expect to see in your code:

Source...
````javascript
                    <Handle
                        key={handleId}
                        type="source" //This is the handle type
                        position={Position.Right} //Just changed this as it would be good to be in parallel with the target
                        id={handleId}
                        style={{ top: positionHandle(i + 1) }}
                    />
````

Target...
````javascript
                    <Handle
                        key={handleId}
                        type="target"  //This is the handle type
                        position={Position.Left}
                        id={handleId}
                        style={{ top: positionHandle(i + 1) }}
                    />
````

If you then try to connect these two together you shouldn't be getting any errors (unless I'm missing something else)... Let me know if that helps out

**Comment by Ipicky22** [comment]
Hey thanks for answer !
I managed to get what I wanted, for the custom node : 

![CustomNodeEnd](https://user-images.githubusercontent.com/37102040/140743646-5f5fe6a2-efda-41dc-814e-7e702983e413.gif)

But impossible to connect it to another node or custom node.

@mermicode, compared to your answer, I have both, handles specified target and others specified source.

```Javascript
const targetHandles = useMemo(
        () =>
            targetArray.map((x: any, i: number) => {
                const handleId = `target-handle-${i + 1}`;
                return (
                    <Handle
                        key={handleId}
                        type="target"
                        position={Position.Left}
                        id={handleId}
                        style={{ top: positionHandle(i + 1) }}
                    />
                );
            }),
        [targetArray, positionHandle]
    );

    const sourceHandles = useMemo(
        () =>
            sourceArray.map((x: any, i: number) => {
                const handleId = `source-handle-${i + 1}`;
                return (
                    <Handle
                        key={handleId}
                        type="source"
                        position={Position.Right}
                        id={handleId}
                        style={{ top: positionHandle(i + 1) }}
                    />
                );
            }),
        [sourceArray, positionHandle]
    );
```

And I return : 

```Javascript
<div ref={nodeRef} >
            {targetHandles}
            <div style={{ display: "flex", justifyContent: "space-around", alignItems: "center" }}>
                <div style={{ flex: 1 }}>
                    <ButtonGroup
                        orientation="vertical"
                        aria-label="vertical outlined button group"
                        size="small"
                    >
                        <Button key="targetMore" onClick={() => add("T")}>+</Button>
                        <Button key="targetLess" onClick={() => less("T")}>-</Button>
                    </ButtonGroup>
                </div>
                <div style={{ flex: 2 }}>
                    <div style={{ fontWeight: 500, fontSize: 15 }}>{data.label}</div>
                </div>
                <div style={{ flex: 1 }}>
                    <ButtonGroup
                        orientation="vertical"
                        aria-label="vertical outlined button group"
                        size="small"
                    >
                        <Button key="sourceMore" onClick={() => add("S")}>+</Button>
                        <Button key="sourceLess" onClick={() => less("S")}>-</Button>
                    </ButtonGroup>
                </div>
            </div>
            {sourceHandles}
        </div>
```

But I advanced in my quest for debugging, because I was able to find two behaviors that I know how to reproduce.
The first is when I click on a handle of my custom node, I get the error seen in previous posts.
For this problem I have absolutely no ideas

```Bash
cjs.js:12 Uncaught TypeError: Cannot read properties of null (reading 'find')

The above error occurred in the <ConnectionLine> component: at ConnectionLine
```

The second behavior is when I try to link a node to my custom node, but starting from the handle of the standart node.

```Bash
couldn't create edge for source handle id: source-handle-1; edge id: reactflow__edge-node_0source-handle-1-node_1null
```

I looked for and found several issues with this warning (https://github.com/wbkd/react-flow/issues/581, https://github.com/wbkd/react-flow/issues/1039, https://github.com/wbkd/react-flow/issues/749), I tried the proposed solutions but without success.

The strangest thing is that when I get my graph, the edge is present. But the targetHandle is null, which corresponds to the handle of the standart node.

![Capture](https://user-images.githubusercontent.com/37102040/140745445-10965b99-80fa-4b87-97ec-c3170d40931e.PNG)

If you have any ideas, I'm interested ^^

**Comment by moklick** [comment]
Hey @Ipicky22 

whenever you update the handles programatically you need to use the `useUpdateNodeInternals` hook to inform ReactFlow about the changes you made https://reactflow.dev/docs/api/hooks/#useupdatenodeinternals

**Comment by Ipicky22** [comment]
Hi @moklick !

Thanks for your response, it's awesome! She solves everything!

So I did :

```Javascript
const nodeId: string = props.id
...
useEffect(() => {
        updateNodeInternals(nodeId)
}, [sourceArray, targetArray])
```
Thank you for the documentation, although I looked for it in the issues, I had not found it.

But thanks to the word "updateNodeInternals" I found this issue ([1038](https://github.com/wbkd/react-flow/issues/1038)) which solves my problem as well ...

Thanks again, and keep it up, ReactFlow is awesome.
